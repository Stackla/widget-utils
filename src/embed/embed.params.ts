import { sanitizeHtmlBasic } from "../libs"

export const generateDataHTMLStringByParams = (params: Record<string, string | boolean | number>): string => {
  const data = Object.entries(params)
    .map(([key, value]) => ` data-${key}="${value}"`)
    .join("")

  return sanitizeHtmlBasic(data)
}
