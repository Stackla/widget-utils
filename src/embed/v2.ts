import { STAGING_LEGACY_WIDGET_URL, PRODUCTION_LEGACY_WIDGET_URL } from "../constants"
import { Environment } from "."
import { generateDataHTMLStringByParams } from "./embed.params"

const getUrlByEnv = (environment: Environment) => {
  switch (environment) {
    case "staging":
      return STAGING_LEGACY_WIDGET_URL
    case "production":
    default:
      return PRODUCTION_LEGACY_WIDGET_URL
  }
}

const getWidgetV2EmbedCode = (data: Record<string, string | boolean | number>) => {
  const dataParams = generateDataHTMLStringByParams(data)

  return `
    <!-- Nosto Widget Embed Code (start) -->
    <div class="stackla-widget" style="width: 100%; overflow: hidden;"${dataParams}></div>
    <!-- Nosto Widget Embed Code (end) -->
    `
}

const invokeV2Javascript = (environment: Environment) => {
  ;(function (d, id) {
    const el = d.scripts[d.scripts.length - 1].previousElementSibling as HTMLDivElement
    if (el) el.dataset.initTimestamp = new Date().getTime().toString()
    const t = d.createElement("script")
    t.src = `${getUrlByEnv(environment)}/media/js/widget/fluid-embed.min.js`
    t.id = id
    ;(d.getElementsByTagName("head")[0] || d.getElementsByTagName("body")[0]).appendChild(t)
  })(document, "stackla-widget-js")
}
export { getWidgetV2EmbedCode, invokeV2Javascript }
