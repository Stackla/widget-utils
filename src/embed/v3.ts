import { STAGING_UI_URL, PRODUCTION_UI_URL } from "../constants"
import { Environment } from "."
import { generateDataHTMLStringByParams } from "./embed.params"

const getUrlByEnv = (environment: Environment) => {
  switch (environment) {
    case "staging":
      return STAGING_UI_URL
    case "production":
    default:
      return PRODUCTION_UI_URL
  }
}

const getWidgetV3EmbedCode = (data: Record<string, string | boolean | number>) => {
  const dataParams = generateDataHTMLStringByParams(data)

  return `<div id="ugc-widget"${dataParams}></div>`
}

const invokeV3Javascript = async (environment: Environment) => {
  const widget = await import(`${getUrlByEnv(environment)}/core.esm.js`)
  widget.init()
}

export { getWidgetV3EmbedCode, invokeV3Javascript }
