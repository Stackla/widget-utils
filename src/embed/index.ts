import {
  STAGING_DATA_URL,
  PRODUCTION_DATA_URL,
  STAGING_LEGACY_WIDGET_DOMAIN,
  PRODUCTION_LEGACY_WIDGET_DOMAIN
} from "../constants"
import { getWidgetV2EmbedCode, invokeV2Javascript } from "./v2"
import { getWidgetV3EmbedCode, invokeV3Javascript } from "./v3"

export type Environment = "staging" | "production"
type Generation = 2 | 3

interface EmbedOptions<T> {
  widgetId: string
  root: T
  environment: Environment
  version?: Generation
  dataProperties: Record<string, string | number | boolean>
}

interface JSONSchema {
  version: number
}

export function getWidgetDataUrl(env: Environment) {
  switch (env) {
    case "staging":
      return STAGING_DATA_URL
    default:
      return PRODUCTION_DATA_URL
  }
}

export function getLegacyWidgetDomain(env: Environment) {
  switch (env) {
    case "staging":
      return STAGING_LEGACY_WIDGET_DOMAIN
    default:
      return PRODUCTION_LEGACY_WIDGET_DOMAIN
  }
}

function getRequestUrl(widgetId: string, environment: Environment) {
  return `${getWidgetDataUrl(environment)}/widgets/${widgetId}/version`
}

async function retrieveWidgetVersionFromServer(widgetId: string, environment: Environment): Promise<number> {
  const response = await fetch(getRequestUrl(widgetId, environment))
  const json: JSONSchema = await response.json()

  return json.version
}

function injectHTML(root: HTMLElement | ShadowRoot, html: string) {
  root.appendChild(document.createRange().createContextualFragment(html))
}

export async function embed<T extends ShadowRoot | HTMLElement>(options: EmbedOptions<T>) {
  const { environment, widgetId, root, version, dataProperties } = options

  try {
    const widgetVersion = version ?? (await retrieveWidgetVersionFromServer(widgetId, environment))
    switch (widgetVersion) {
      case 2:
        window.stackWidgetDomain = getLegacyWidgetDomain(environment)
        dataProperties.hash = widgetId
        injectHTML(root, getWidgetV2EmbedCode(dataProperties))
        invokeV2Javascript(environment)
        break
      case 3:
        dataProperties.wid = widgetId
        injectHTML(root, getWidgetV3EmbedCode(dataProperties))
        await invokeV3Javascript(environment, widgetId)
        break
      default:
        throw new Error(`No widget code accessible with version ${widgetVersion}`)
    }
  } catch (error) {
    console.error(`Failed to embed widget. ${error}`)
  }
}
