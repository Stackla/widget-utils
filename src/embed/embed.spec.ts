/* eslint-disable no-var */
import { embed } from "."
import fetchMock from "jest-fetch-mock"
import { getWidgetV2EmbedCode } from "./v2"
import { getWidgetV3EmbedCode } from "./v3"
import { generateDataHTMLStringByParams } from "./embed.params"

fetchMock.enableMocks()

const REQUEST_URL = "https://widget-data.stackla.com/widgets/123/version"
const STAGING_REQUEST_URL = "https://widget-data.teaser.stackla.com/widgets/123/version"

describe("load embed code", () => {
  beforeEach(() => {
    fetchMock.resetMocks()
  })

  it("should return the correct embed code for v2", async () => {
    fetchMock.mockIf(REQUEST_URL, async () => {
      return JSON.stringify({ version: 2 })
    })

    const createdDiv = document.createElement("div")
    await embed({
      widgetId: "123",
      root: createdDiv,
      dataProperties: {
        foo: "bar",
        baz: 123
      },
      environment: "production"
    })

    expect(createdDiv.innerHTML).toContain(getWidgetV2EmbedCode({ foo: "bar", baz: 123, hash: "123" }))
    expect(createdDiv.innerHTML).toContain("var t, el = d.scripts[d.scripts.length - 1].previousElementSibling;")
  })

  it("should test staging for v2", async () => {
    fetchMock.mockIf(STAGING_REQUEST_URL, async () => {
      return JSON.stringify({ version: 2 })
    })

    const createdDiv = document.createElement("div")
    await embed({
      widgetId: "123",
      root: createdDiv,
      dataProperties: {
        foo: "bar",
        baz: 123
      },
      environment: "staging"
    })

    expect(window.stackWidgetDomain).toBe("widgetapp.teaser.stackla.com")
  })

  it("should test production for v2", async () => {
    fetchMock.mockIf(REQUEST_URL, async () => {
      return JSON.stringify({ version: 2 })
    })

    const createdDiv = document.createElement("div")
    await embed({
      widgetId: "123",
      root: createdDiv,
      dataProperties: {
        foo: "bar",
        baz: 123
      },
      environment: "production"
    })

    expect(window.stackWidgetDomain).toBe("widgetapp.stackla.com")
  })

  it("should return the correct embed code for v3", async () => {
    fetchMock.mockIf(REQUEST_URL, async () => {
      return JSON.stringify({ version: 3 })
    })

    const createdDiv = document.createElement("div")
    await embed({
      widgetId: "123",
      root: createdDiv,
      dataProperties: {
        foo: "bar",
        baz: 123
      },
      environment: "production"
    })

    expect(createdDiv.innerHTML).toContain(getWidgetV3EmbedCode({ foo: "bar", baz: 123, wid: "123" }))
    expect(createdDiv.innerHTML).toContain(`const widget = await import('https://widget-ui.stackla.com/core.esm.js');`)
  })

  it("should throw an error if the version is not supported", async () => {
    fetchMock.mockIf(REQUEST_URL, async () => {
      return JSON.stringify({ version: 4 })
    })

    const createdDiv = document.createElement("div")
    try {
      await embed({
        widgetId: "123",
        root: createdDiv,
        dataProperties: {
          foo: "bar",
          baz: 123
        },
        environment: "production"
      })
    } catch (error) {
      expect(error).toBe("Failed to embed widget. No widget code accessible with version 4")
    }
  })

  it("should skip the fetch call if the version is provided", async () => {
    const createdDiv = document.createElement("div")
    await embed({
      widgetId: "123",
      root: createdDiv,
      version: 3,
      dataProperties: {
        foo: "bar",
        baz: 123
      },
      environment: "production"
    })

    expect(fetchMock).not.toHaveBeenCalled()
    expect(createdDiv.innerHTML).toContain(getWidgetV3EmbedCode({ foo: "bar", baz: 123, wid: "123" }))
    expect(createdDiv.innerHTML).toContain("const widget = await import('https://widget-ui.stackla.com/core.esm.js');")
  })

  it("should test param string method", async () => {
    const params = generateDataHTMLStringByParams({ foo: "bar", baz: 123, wid: "123" })

    expect(params).toBe(' data-foo="bar" data-baz="123" data-wid="123"')
  })

  it("should deal with malicious payloads", async () => {
    const createdDiv = document.createElement("div")
    await embed({
      widgetId: "123",
      root: createdDiv,
      version: 3,
      dataProperties: {
        foo: "bar",
        baz: 123,
        '><img src="x" onerror="alert(1)">': '"><img src="x" onerror="alert(1)">'
      },
      environment: "production"
    })

    expect(fetchMock).not.toHaveBeenCalled()
    expect(createdDiv.innerHTML).toContain(
      `<div id="ugc-widget" data-foo="bar" data-baz="123" data-%3e%3cimg%20src%3d%22x%22%20onerror%3d%22alert(1)%22%3e="%22%3E%3Cimg%20src%3D%22x%22%20onerror%3D%22alert(1)%22%3E" data-wid="123"></div>`
    )
  })
})
