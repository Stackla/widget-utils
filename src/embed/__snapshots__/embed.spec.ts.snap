// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`load embed code should return the correct embed code for v2 1`] = `
"
    <!-- Nosto Widget Embed Code (start) -->
    <div class="stackla-widget" style="width: 100%; overflow: hidden;" data-foo="bar" data-baz="123" data-hash="123"></div>
    <!-- Nosto Widget Embed Code (end) -->
    "
`;
